feat: Adds progress bar to cloud storage downloads

Linus045 opened the pull request:
~~This is currently a **draft PR**.
I just wanted to get general feedback on the whole idea and implementation, as I'm quite new to this code base.~~

The goal of this PR is to provide a progress bar when downloading files via cloud storage e.g. Dropbox/WebDav/FTP.
In the current implementation this only works for Dropbox and WebDav, **for FTP the file size is not available when downloading**.

Depending on the use case it might be best to move this "progress bar window"-component into it's own widget, so it can be used for other things as well (e.g. #11217).

Here is a picture of the new download window:
<img src="https://github.com/user-attachments/assets/7eac4813-2401-4b66-aa59-0674252bac55" width="250"> <img src="https://github.com/user-attachments/assets/774be9c4-2912-4057-bc4e-26b0af3da095" width="250">

Here is the current status with FTP. The filesize is 'N/A' and we don't show the progress bar in the dialog:
<img src="https://github.com/user-attachments/assets/ac2a773d-13fa-4b47-bc6c-4e2101063182" width="250"><img src="https://github.com/user-attachments/assets/6cd73eff-6911-4998-a66d-62e5b0ed5b66" width="250">

Stuff that needs more consideration:
- [ ] ~~How often should the progress bar update (to limit redraw calls)? currently this is set to update every 20% progress increase (so at 20%, 40%,..)~~
- [ ] I'm not that familiar with the "luasocket" so the file sink wrapper might need some changes especially in case errors happen, I'm not sure how to test/debug  this
- [ ] ~~variable naming and code structuring (see TODO's in code)~~
- [ ] In general more testing is needed, to verify that this feature does not cause any crashes


Regarding FTP downloads, it might be possible to request the file size before we start the actual download so we can show a progress bar as well, however this should be done in a new separate issue.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/13650)
<!-- Reviewable:end -->

hius07 commented:
Thanks, I think it's interesting.
Agree that the progress window should be a separate widget.
No need to build a mighty TitleBar, it can be just two strings (TextWidget).

hius07 commented:
I think the redraw step should depend on a file size.
Downloading a small file, no need to redraw five times within 1-2 seconds.

Frenzie commented:
> How often should the progress bar update (to limit redraw calls)? currently this is set to update every 20% progress increase (so at 20%, 40%,..)

I'd also (or perhaps only) add a time aspect to it. See https://koreader.rocks/doc/modules/time.html for related functions.

Linus045 commented:
I've extracted the progress window into its own dialog component and replaced the TitleBar with TextWidgets.

I've added 3 different refresh modes which determine when the redraw occurs, this might be useful for other use cases later on.
- "time" uses `os.time()` and refreshes every x seconds (lower precision not possible and I would argue not needed)
- "fixed_percentage" and "steps" update at certain progress steps
- "steps" also shows little markers (might be useful if something differnt than a download progress needs to be show?)

refresh_mode - "time"/"fixed_percentage"
<img src="https://github.com/user-attachments/assets/93f85feb-3912-4138-98dc-a5d5b6aec504" width="250">

refresh_mode - "steps"
<img src="https://github.com/user-attachments/assets/a19f3652-c897-4f2a-9305-2f6a72235568" width="250">

When there is no `max_progress` the progress bar will be hidden, this is currently the case for FTP downloads:
<img src="https://github.com/user-attachments/assets/0dc82d4e-440a-49f8-b4d7-8a0c035ff8bd" width="250">



These are mostly just some ideas on how to flesh out the progress dialog.
If you think these are unnecessary I'm happy to remove them.


About the ability to abort the download, I'm not sure how I would implement the trapper but I like the the iidea. 
Although I think this might be something for a follow up issue, as the code changes are probably going to be significant as well.

hius07 commented:
> I'm not sure how I would implement the trapper but I like the the iidea.

For example:
https://github.com/koreader/koreader/blob/24f1a80ce8571e50356bd970792218f1c118492c/plugins/opds.koplugin/opdsbrowser.lua#L1065-L1070
https://github.com/koreader/koreader/blob/24f1a80ce8571e50356bd970792218f1c118492c/plugins/opds.koplugin/opdsbrowser.lua#L1102-L1111

Frenzie commented:
I don't really see the point for anything other than time. The other settings need time as a factor regardless to prevent sending a crazy amount of screen updates, and conversely when they don't you'd be staring at no progress while in reality there might've been 1% progress.

I suppose maybe in case there are a few distinct steps more so than "progress".

Linus045 commented:
I would agree that in this specific case for downloading files "time" might be the only useful mode 

> The other settings need time as a factor regardless to prevent sending a crazy amount of screen updates

I would disagree, in the "steps" and "fixed_percentage" mode the screen is only updated once a significant change occurred, so unless you set the refresh value to something like every 1% it should be fine. 
Although  I can see what you mean when we're  downloading a small file, the progress would happen so quickly that we would force unnecessary redraw calls.
However I indented these two modes ("steps"/"fixed_percentage") for different use cases - when no exact progress can be calculated.

I not sure of a good use case yet, however I could image a case where we need to do modifications on a lot of files and we wanna show the progress update on a per file basis. E.g. deleting 100 files or something.

Though in general I added the other modes mostly for future use. I'm fine with removing them if there will be no real need for them in the future.

Frenzie commented:
> I would disagree, in the "steps" and "fixed_percentage" mode the screen is only updated once a significant change occurred, so unless you set the refresh value to something like every 1% it should be fine.

You left off the second part of the sentence. You can either use always right or you can use always wrong to some degree but accidentally more or less right in some edge cases (aka having the exact same hardware as the developer).

> Though in general I added the other modes mostly for future use

On average that's a terrible reason to add anything. ;-)

Linus045 commented:
> You left off the second part of the sentence. You can either use always right or you can use always wrong to some degree but accidentally more or less right in some edge cases (aka having the exact same hardware as the developer).

I'm sorry but I don't understand what you're trying to say here, could you clarify please?

> On average that's a terrible reason to add anything. ;-)

Yeah after a break and some consideration I agree that the other modes don't belong in this PR. They can be added later on in new PRs if needed.
I'll remove the modes and just update based on elapsed time to prevent unnecessary redraws.


Regarding the Trapper, I would like to do that in a new Issue/PR instead. I'll create an Issue once this PR has been resolved.

Frenzie commented:
> I'm sorry but I don't understand what you're trying to say here, could you clarify please?

I mean that hardware speeds are all over the place.

hius07 commented on frontend/ui/widget/progressbardialog.lua:
```diff
@@ -0,0 +1,191 @@
+    self.dimen = Screen:getSize()
+
+    -- set default values if not provided
+    self.title = self.title ~= nil and self.title or nil
```
I don't get, what does this line do?

Linus045 replied:
I had a different default value while testing and ended up changing the default to nil anyway. 

I left it in there to show the intention that title, subtile and progress_max was nil-able but at this point it should can be obvious so this can be removed, the comments and usage make this clear enough

hius07 commented on frontend/ui/widget/progressbardialog.lua:
```diff
@@ -0,0 +1,191 @@
+    local progress_bar_width = Screen:scaleBySize(360)
+    local progress_bar_height = Screen:scaleBySize(18)
+
+    self.title_text = TextWidget:new {
```
Can this be moved into the `if` below?

hius07 commented:
I think the centered content is good.

Please order the `require` section (in alphabetical order, case sensitive, see other widgets for example).

I'd propose to have the defaults
```Lua
local ProgressbarDialog = WidgetContainer:extend{
    title = _("Downloading…"),
    refresh_time_seconds = 3,
}

```
and remove them from the caller.

All looks good (not tested).

Linus045 commented:
I would agree to move `refresh_time_seconds` into the `extend{}` block as default. 

However for the `title` i would disagree.
If this widget is used for other purposes (e.g. file processing etc.) later on (which was the reason to move it into its own widget component in the first place)  it would require a different title (i.e. "Processing...") and we would need to overwrite it on the caller-side anyway. 
In that case the default title ("Downloading...") would only be used for the file downloading in cloud storage and in most other cases we would overwrite it (or maybe even explicitly set it to nil to hide the title).

**Basically** it comes down to this question: "Is downloading files the default use case for most progress dialogs?" to which I would say no.
IMO it's better to require the caller to actively make the decision of what the title (and subtitle) should be.

(Also yes this discussion is a bit nit-picky haha)

Linus045 commented on frontend/ui/widget/progressbardialog.lua:
```diff
@@ -0,0 +1,178 @@
+    if self.subtitle then
+        vertical_group[#vertical_group + 1] = TextWidget:new {
+            text = self.subtitle or "",
+            face = Font:getFace("cfont", 16),
```
The predefined fonts work just fine. I didn't look too much into the font stuff and wasn't aware they existed.
